### PR TITLE
XL/objects: Initialize format.json outside of erasure.

### LINKED ISSUE #1636: XL: data corruption/loss may happen if underlying disks are renamed - May 20th
Its known that physical disks may named different on each run of operating system especially in GNU/Linux, eg. a device `sdi` may appear as `sdk` in next reboot.

In a system where UUID of disk file system is not used in fstab or mount command to mount such disks may lead XL to place data in a wrong set of disk order.  This will lead to data corruption and loss.

This needs to be fixed.

### PR DESCRIPTION
**Pull request by harshavardhana**
Fixes #1636

New format now generates a UUID and includes it along with
the order of disks. So that UUID is the real order of disks
and on command line user is able to specify disks in any order.

This pre-dominantly solves our dilemma.

```
{
   "uuid" : "00e4cf06-5bf5-4bb5-b885-4b2fff4a7959",
   "format" : "xl",
   "xl" : {
      "disksUUIDMap" : {
         "/tmp/bc8" : "c47d2608-5067-4ed7-b1e4-fb81bdbb549f",
         "/tmp/bc1" : "00e4cf06-5bf5-4bb5-b885-4b2fff4a7959",
         ....
         "/tmp/bc2" : "b21bbcef-b1ac-490e-8005-2e742c04a9cb",
         "/tmp/bc6" : "2f04d741-869e-498d-9f5a-9feefc92e486",
         "/tmp/bc4" : "ed60fa5c-4ee4-4409-a526-05f6c02bd84c",
         "/tmp/bc5" : "adbabad0-3e0e-43f1-84ba-da5b64e42d60"
      },
      "version" : "1",
      "disks" : [
         "00e4cf06-5bf5-4bb5-b885-4b2fff4a7959",
         ....
         "c47d2608-5067-4ed7-b1e4-fb81bdbb549f",
         "a543293e-99f1-4310-b540-1e450878e844",
         "18f97cbe-529a-456a-b6d4-0feacf64534d"
      ]
   },
   "version" : "1"
}
```

### CODE REVIEW

**Comment by harshavardhana** [comment]
Available for review, comments haven't been added yet. Before we finalize on the format but it is working pretty well without sticking to the order of disks on command line.

**Comment by abperiasamy** [comment]
{
   "format" : "xl",
   "xl" : {
      "version" : "1",
      "disk" : "00e4cf06-5bf5-4bb5-b885-4b2fff4a7959",
      "jbod" : [
         "00e4cf06-5bf5-4bb5-b885-4b2fff4a7959",
         ....
         "c47d2608-5067-4ed7-b1e4-fb81bdbb549f",
         "a543293e-99f1-4310-b540-1e450878e844",
         "18f97cbe-529a-456a-b6d4-0feacf64534d"
      ]
   },
   "version" : "1"
}

xlLoadFormatJSON(dirs []string) dirs []string

File: format-config-v1.go [hunk 1]
```diff
@@ -37,66 +44,104 @@ type formatConfigV1 struct {
...
+		}
+		// Validate format version and format type.
+		if formatXL.Version != "1" {
+			return nil, fmt.Errorf("Unsupported version of backend format [%s] found.", formatXL.Version)
+		}
+		if formatXL.Format != "xl" {
+			return nil, fmt.Errorf("Unsupported backend format [%s] found.", formatXL.Format)
+		}
+		// Save valid formats in a map.
+		jbodMap[disk] = formatXL
```
Comment (abperiasamy) [inline format-config-v1.go]: This call is directly reading from the disk. It is supposed to create a boostrap stack based on disk type (network or local) and then load it.

File: format-config-v1.go [hunk 2]
```diff
@@ -37,66 +44,104 @@ type formatConfigV1 struct {
...
+	return jbodMap, nil
 }
 
-// checkFormat - validates if format.json file exists.
-func checkFormat(storage StorageAPI) error {
-	_, err := storage.StatFile(minioMetaBucket, formatConfigFile)
-	if err != nil {
-		return err
+// checkFormatXL - verifies if format.json file exists and is accessible.
+func checkFormatXL(disks ...string) error {
```
Comment (abperiasamy) [inline format-config-v1.go]: func checkFormatXL(jbodMap map[string]formatConfigV1) error {

check format should check for consistencies instead. (Format, version and JBOD UUIDs)

File: format-config-v1.go [hunk 3]
```diff
@@ -37,66 +43,191 @@ type formatConfigV1 struct {
 	XL      *xlFormat `json:"xl,omitempty"`
 }
 
-// FIXME: currently we don't check single exportPath which uses FS layer.
+// validateXLJBOD - validate xl jbod order if they are consistent.
+func validateXLJBOD(jbod []string, jsonMap map[int]formatConfigV1) error {
+	jbodStr := strings.Join(jbod, ", ")
+	for _, format := range jsonMap {
+		if format.XL == nil {
+			continue
```
Comment (abperiasamy) [inline format-config-v1.go]: Add comment here: This disk is offline. OK to ignore and continue.

File: format-config-v1.go [hunk 4]
```diff
@@ -37,66 +43,191 @@ type formatConfigV1 struct {
...
+		}
 	}
-	if formatXL.Format != "xl" {
-		return nil, fmt.Errorf("Unsupported backend format [%s] found.", formatXL.Format)
+	return newDisks, nil
+}
+
+// loadFormatXL - load XL format.json.
+func loadFormatXL(storageDisks ...StorageAPI) (jbodMap map[int]formatConfigV1, err error) {
+	var fileNotFoundCount = 0
```
Comment (abperiasamy) [inline format-config-v1.go]: func loadFormatXL(storageDisks ...StorageAPI) (formatConfigs map[string]formatConfigV1, err error) {

where index is disk

Comment (harshavardhana) [inline format-config-v1.go]: > func loadFormatXL(storageDisks ...StorageAPI) (formatConfigs map[string]formatConfigV1, err error) {
> 
> where index is disk

disk is a string, i avoided that by not passing in the disks argument. So kept the index which is also correct. Just indicates the index of first disk, second disk ..

File: format-config-v1.go [hunk 5]
```diff
@@ -37,66 +43,191 @@ type formatConfigV1 struct {
...
+	return jbodMap, nil
 }
 
-// checkFormat - validates if format.json file exists.
-func checkFormat(storage StorageAPI) error {
-	_, err := storage.StatFile(minioMetaBucket, formatConfigFile)
-	if err != nil {
-		return err
+// checkFormatXL - verifies if format.json format is intact.
+func checkFormatXL(disks []string, jbodMap map[int]formatConfigV1) error {
```
Comment (abperiasamy) [inline format-config-v1.go]: Reduce two arguments into one. jbodMap map[string]formatConfigV1)

**Comment by harshavardhana** [comment]
> Reduce two arguments into one. jbodMap map[string]formatConfigV1)

It is needed because i wanted to verify if the disks argument passed from command line has equal amount of disks in JBOD.

File: format-config-v1.go [hunk 6]
```diff
@@ -37,66 +43,191 @@ type formatConfigV1 struct {
...
-	w, err := storage.CreateFile(minioMetaBucket, formatConfigFile)
-	if err != nil {
-		return err
-	}
-	formatXL := formatConfigV1{
-		Version: "1",
-		Format:  "xl",
-		XL:      xl,
+// saveFormatXL - save XL format configuration on all disks.
+func saveFormatXL(storageDisks ...StorageAPI) (jbodMap map[int]formatConfigV1, err error) {
```
Comment (abperiasamy) [inline format-config-v1.go]: Rename to initFormatXL

File: format-config-v1.go [hunk 7]
```diff
@@ -37,66 +43,191 @@ type formatConfigV1 struct {
...
+}
 
-// loadFormatXL - load XL format.json.
-func loadFormatXL(storage StorageAPI) (xl *xlFormat, err error) {
-	offset := int64(0)
-	r, err := storage.ReadFile(minioMetaBucket, formatConfigFile, offset)
-	if err != nil {
-		return nil, err
+// arrangeDiskInOrder - arrange disks in saved order.
+func arrangeDiskInOrder(storageDisks []StorageAPI, jbodMap map[int]formatConfigV1) ([]StorageAPI, error) {
```
Comment (abperiasamy) [inline format-config-v1.go]: Rename to reorderDisks

File: format-config-v1.go [hunk 8]
```diff
@@ -37,66 +43,191 @@ type formatConfigV1 struct {
...
+			savedJBOD = format.XL.JBOD
+			break
+		}
 	}
-	decoder := json.NewDecoder(r)
-	formatXL := formatConfigV1{}
-	err = decoder.Decode(&formatXL)
-	if err != nil {
+	if err := validateXLJBOD(savedJBOD, jbodMap); err != nil {
 		return nil, err
```
Comment (abperiasamy) [inline format-config-v1.go]: This check should be performed in checkFormatXL.

File: format-config-v1.go [hunk 9]
```diff
@@ -37,66 +43,191 @@ type formatConfigV1 struct {
 	XL      *xlFormat `json:"xl,omitempty"`
 }
 
-// FIXME: currently we don't check single exportPath which uses FS layer.
+// validateXLJBOD - validate xl jbod order if they are consistent.
+func validateXLJBOD(jbod []string, jsonMap map[int]formatConfigV1) error {
```
Comment (abperiasamy) [inline format-config-v1.go]: func checkJBODConsistencyj(formatConfigs map[string]formatConfigV1) error {

File: format-config-v1.go [hunk 10]
```diff
@@ -37,66 +43,191 @@ type formatConfigV1 struct {
...
+	jbodStr := strings.Join(jbod, ", ")
+	for _, format := range jsonMap {
+		if format.XL == nil {
```
Comment (abperiasamy) [inline format-config-v1.go]: let format itself be set to nil for missing or uninitialized disks.

File: format-config-v1.go [hunk 11]
```diff
@@ -37,66 +43,191 @@ type formatConfigV1 struct {
...
-func loadFormatXL(storage StorageAPI) (xl *xlFormat, err error) {
-	offset := int64(0)
-	r, err := storage.ReadFile(minioMetaBucket, formatConfigFile, offset)
-	if err != nil {
-		return nil, err
+// arrangeDiskInOrder - arrange disks in saved order.
+func arrangeDiskInOrder(storageDisks []StorageAPI, jbodMap map[int]formatConfigV1) ([]StorageAPI, error) {
+	var savedJBOD []string
+	for index := range storageDisks {
+		if format, ok := jbodMap[index]; ok {
```
Comment (abperiasamy) [inline format-config-v1.go]: Change this logic to simply deepEquals first JBOD to the rest. If any fails, the whole check fails.

File: format-config-v1.go [hunk 12]
```diff
@@ -37,66 +43,191 @@ type formatConfigV1 struct {
...
+			if format.XL == nil {
```
Comment (abperiasamy) [inline format-config-v1.go]: let format itself be set to nil for missing or uninitialized disks.

File: format-config-v1.go [hunk 13]
```diff
@@ -37,66 +43,191 @@ type formatConfigV1 struct {
...
-// FIXME: currently we don't check single exportPath which uses FS layer.
+// validateXLJBOD - validate xl jbod order if they are consistent.
+func validateXLJBOD(jbod []string, jsonMap map[int]formatConfigV1) error {
+	jbodStr := strings.Join(jbod, ", ")
+	for _, format := range jsonMap {
+		if format.XL == nil {
+			continue
+		}
+		savedJBODStr := strings.Join(format.XL.JBOD, ", ")
+		if jbodStr != savedJBODStr {
```
Comment (abperiasamy) [inline format-config-v1.go:63]: Simply deepEqual the rest of JBODs to the first.

File: format-config-v1.go [hunk 14]
```diff
@@ -37,66 +43,191 @@ type formatConfigV1 struct {
...
+			return errors.New("Inconsistent disks.")
+		}
```
Comment (abperiasamy) [inline format-config-v1.go:65]: Tell which disk is inconsistent with what.

**Comment by harshavardhana** [comment]
> Simply deepEqual the rest of JBODs to the first.

I thought it is better to do it string style.

**Pull request merged**